rfcs/RFC-Uploads.md

```diff
@@ -0,0 +1,348 @@
+- DP(Data Portal) - The cellxgene Data Portal.
+- Small File - a file < 5 GB in size.
+- Large File - a file > 5 GB in size.
+
+## Problem Statement | Background
+
+One of the primary goals of the DP is to allow users to upload their own data sets. These data sets can vary in
+size from a few MB to several GB. User may want to upload from their local machine or from their CSP. For
+users that want to upload from their local machine, they must be able to resume an upload in the event their connection
+is interrupted. For users that want to upload data sets from their CSP, they are able to provide a shared link from
+their CSP that we can used to upload directly to our storage.
+
+## Product Requirements
+
+1. A user can upload a file to DP using a shared link from their CSP.
+1. A user can upload a file from their local machine using the DP browser.
+1. A user can resume an upload with out completely restarting.
+1. The user must be authenticated before uploading.
+
+### Nonfunctional requirements
+
+Nonfunctional requirements are not directly related to the user story but may instead reflect some technical aspect of the implementation, such as latency. [Here's](https://en.wikipedia.org/wiki/Non-functional_requirement) a more comprehensive list of nonfunctional requirements.
+
+1. Cloud to Cloud uploads will retry if they fail.
+1. A user can upload a single file up to 30GB in size.
+
+## Detailed Design | Architecture | Implementation
+
+![Upload Design](https://app.lucidchart.com/publicSegments/view/a74f747f-d071-41ea-949f-52f38a98747a/image.png)
+Figure 1: The Upload Design
+
+### Data Portal Browser App
+
+The DP Browser App will be used authenticate the user, and interact with the DP Backend to upload files.
+It will also use AWS JS SDK to upload files from local storage to AWS. These feature will extend the existing DP
+Frontend.
+
+### Data Portal Backend
+
+The DP Backend is responsible for handling users requests. It will generated presigned URL and access tokens
+for the DP Browser App to facilitate the uploading of files. It will also enqueue shared links for downloading.
+This additional functionality can be built into the existing DP Backend Lambda.
+
+### Download Queue
+
+The Download Queue is an [AWS SQS](https://aws.amazon.com/sqs/) that will track pending and in progress download jobs.
+
+#### Download Job Entry
+
+These are the fields that will be in the download job placed in the queue:
+
+- Link - a to the link to download. Use to retrieve the file from CSP.
+- collect_id - identifies the collection. Used to determine the storage location.
+- dataset_id - identifies the dataset. Used to determine the storage location
+- asset_id - identifies the asset. Used to determine the storage location, and file name.
+- file size - the size of the asset to download. Used to determine the download method, and EC2 instance size.
+
+### Data Portal APIs
+
```

Comment:
Perhaps replace the following two sentences with:

These APIs replace the `PUT /v1/submission/{project_uuid}/file` section and references in the Corpora High Level Architecture.

---


 Is the intent to merge this RFC Into that design document for clarity?

There are some details described in the original API that are missing here such as:

> A quick validation will be performed to ensure that if the file is a data file, then the extension is one of the accepted types (i.e. .loom, .h5ad, or .rds). If a file already exists with the same name, it will replace it and increment the revision value of the dataset.

The responses below only document success cases and not failures such as:

> 400 Invalid filename or filetype: {status, message}
> 401 Unauthorized: {status, message}
> 404 Submission not found: {status, message}